packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -100,7 +100,7 @@ const TextHook: Hooks<VText> = {
                 assert.fail('Hydration mismatch: incorrect node type received.');
             }
 
-            if (node.nodeValue !== vNode.text) {
+            if (node.nodeValue !== String(vNode.text)) {
```

Comment:
I was actually looking at this recently and wondering if it wouldn't make sense to cast `vNode.text` to a string earlier in the process. At some point we will call `renderer.createText(vnode.text)`, which calls `document.createTextNode()`, which casts integers/booleans/etc to a string.

E.g. we could cast it here:

https://github.com/salesforce/lwc/blob/647de31f30be1084253a85b8b850fb72de7294ac/packages/%40lwc/engine-core/src/framework/api.ts#L673-L676

(There are probably other places we would have to do this as well.)

Reply:
I'm open to doing that change. However, I prefer to leave things as-is for the engine, mainly because in the case you described, the browser is the one casting at the time of rendering, and we need to do this cast for the hydration only during dev as a validation.

Reply:
Ah, that's a good point. I guess I assumed if we casted it to a string earlier on, though, then the JS engine could do further optimizations because the type would be consistent. I didn't test it, though, so I have no evidence for that. 🙂

Reply:
I would also recommend doing the normalization before creating the VNode. 
- It is hard to understand why you are explicitly converting `vNode.text` to a string. Based on the types, this field is already a string or undefined.
- From a pure rendering correctness perspective, we don't want to diff the original object but the string value.

Reply:
done in 670a3a5